streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import java.util.Map;
+
+/**
+ * Branches the records in the original stream based on the predicates supplied for the branch definitions.
+ * <p>
+ * Branches are defined with {@link BranchedKStream#branch(Predicate, Branched)} or
+ * {@link BranchedKStream#defaultBranch(Branched)} methods. Each record is evaluated against the predicates
+ * supplied via {@link Branched} parameters, and is routed to the first branch for which its respective predicate
+ * evaluates to {@code true}.
+ * <p>
+ * Each branch (which is a {@link KStream} instance) then can be processed either by
+ * a {@link java.util.function.Function} or a {@link java.util.function.Consumer} provided via a {@link Branched}
+ * parameter. It also can be accessed from the {@link Map} returned by {@link BranchedKStream#defaultBranch(Branched)} or
+ * {@link BranchedKStream#noDefaultBranch()} (see <a href="#examples">usage examples</a>).
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
+ * to route a record to multiple streams, you can use {@link KStream#filter(Predicate)} for each predicate instead
+ * of branching.
+ * <p>
+ * The process of routing the records to different branches is a stateless record-by-record operation.
+ * <h2><a name="maprules">Rules of forming the resulting map</a></h2>
+ * The keys of the {@code Map<String, KStream<K, V>>} entries returned by {@link BranchedKStream#defaultBranch(Branched)} or
+ * {@link BranchedKStream#noDefaultBranch()} are defined by the following rules:
+ * <p>
+ * <ul>
+ *     <li>If {@link Named} parameter was provided for {@link KStream#split(Named)}, its value is used as
+ *     a prefix for each key. By default, no prefix is used
+ *     <li>If a name is provided for the {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, its value is appended to the prefix to form the {@code Map} key

```

Comment:
Do we use dots at the end of list items? (Sorry -- English is my second language so I really don't know. I noticed that sometimes you do but in most times you don't. In Russian language we always do, but Russian and English punctuation are completely different.)

Reply:
Also my second language... I think, it is mostly a matter of style. Personally, I prefer to use `.` at the end, because a bullet point is still a sentence from my POV, and actually, a bullet point could be multiple sentences. You also capitalized the first word of each bullet point, so it seems to be a sentence. -- I guess some people don't capitalize the first word and also don't use a `.` at the en, ie, us a "no-sentence" bullet point style. But this only work for "single sentence" bullet points.

It's really a nit. I just raised my personal (an obviously very subjective) preference.